ssl/ssl_key_share.cc

```diff
@@ -176,20 +178,550 @@ class X25519KeyShare : public SSLKeyShare {
+      return false;
+    }
+
+    // Initialize the context with the pkey
+    ctx_.reset(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Retrieve the lengths of the ciphertext and shared secret
+    size_t ciphertext_len = 0;
+    size_t secret_len = 0;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), nullptr, &ciphertext_len,
+                              nullptr, &secret_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // EVP_PKEY_encapsulate() will generate a shared secret, then encrypt it
+    // with the peer's public key. We send the resultant ciphertext to the
+    // peer by writing it to |out_public_key|, then...
+    Array<uint8_t> ciphertext;
+    Array<uint8_t> shared_secret;
+    if (!ciphertext.Init(ciphertext_len) || !shared_secret.Init(secret_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+      return false;
+    }
+
+    size_t ciphertext_bytes_written = ciphertext_len;
+    size_t secret_bytes_written = secret_len;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), ciphertext.data(),
+                              &ciphertext_bytes_written, shared_secret.data(),
+                              &secret_bytes_written) ||
+                              ciphertext_bytes_written != ciphertext_len ||
+                              secret_bytes_written != secret_len ||
+                              !CBB_add_bytes(out_public_key, ciphertext.data(), ciphertext_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // ...we retain the shared secret for our own use.
+    *out_secret = std::move(shared_secret);
+
+    // Success; clear the alert
+    *out_alert = 0;
+    return true;
+  }
+
+  // Because this is a KEM key share, |peer_key| is actually the ciphertext
+  // resulting from the peer encapsulating the shared secret under our public key.
+  // In Finish(), we use our previously generated private key to decrypt
+  // that ciphertext and obtain the shared secret.
+  bool Finish(Array<uint8_t> *out_secret, uint8_t *out_alert,
+              Span<const uint8_t> peer_key) override {
+    // Basic nullptr checks
+    if (!out_alert) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // Set alert to internal error by default
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+
+    if (!out_secret) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should have been initialized by a previous call to Offer()
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+      return false;
+    }
+
+    // Retrieve the lengths of the ciphertext and shared secret
+    size_t ciphertext_len = 0;
+    size_t secret_len = 0;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), nullptr, &ciphertext_len, nullptr,
```

Comment:
NP: you can do the same thing with decapsulate and it looks just a little strange to call encapsulate before decapsulate.

Reply:
For encapsulate, `ciphertext_len` is an output of the function. For decapsulate, `ciphertext_len` is an input to the function.

Reply:
Not sure I get it, aren't you calling encapsulate with null to get the size? You can do that with [EVP_PKEY_decapsulate](https://github.com/aws/aws-lc/blob/main/include/openssl/evp.h#L750-L752) too.

Reply:
Yes, [the call to `EVP_PKEY_encapsulate()` with NULL values is setting the sizes of both **`ciphertext_len`** and **`secret_len`**](https://github.com/aws/aws-lc/blob/main/crypto/evp_extra/p_kem.c#L76-L77). 

However the type signatures for `EVP_PKEY_encapsulate()` and `EVP_PKEY_decapsulate()` are different. [Calling `EVP_PKEY_decapsulate()` with NULL values will only set **`shared_secret_len`**.](https://github.com/aws/aws-lc/blob/main/crypto/evp_extra/p_kem.c#L139)

`EVP_PKEY_encapsulate()` has a `size_t *ciphertext_len` so it can modify the callers value of `ciphertext_len` that the pointer is pointing at, whearas `EVP_PKEY_decapsulate()` only has a `size_t ciphertext_len` so any modifications can't be passed up to the caller.

```
int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
                         uint8_t *ciphertext, size_t *ciphertext_len,
                         uint8_t *shared_secret, size_t *shared_secret_len);
```
and 

```
int EVP_PKEY_decapsulate(EVP_PKEY_CTX *ctx,
                         uint8_t *shared_secret, size_t *shared_secret_len,
                         uint8_t *ciphertext, size_t ciphertext_len);
```

Reply:
but that's because for `encapsulate` both the ciphertext and the shared secret are output arguments. For `decapsulate`, the shared secret is an output argument and the ciphertext is an input argument.

Reply:
In this case I don't think you need the ciphertext length, the KEM decapsulate checks this already https://github.com/aws/aws-lc/blob/main/crypto/evp_extra/p_kem.c#L143-L148

Reply:
Made this fix. However, since the size checks now happen inside the decapsulate function instead of before at the SSL layer, the client will now receive `SSL_AD_INTERNAL_ERROR` TLS Alert instead of `SSL_AD_DECODE_ERROR` if they send an improperly sized KEM ciphertext.